src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+# Proposal: QIR Adaptor Tool Specification
+
+This document discuss a tool that transforms a QIR into a restricted version of the QIR (known as a profile).
+We aim to make a specification for a generic tool that allows the user to
+
+1. Create or use an existing profile without the need of writing code.
+2. Validate that a QIR is compliant with the specific profile.
+3. Generate a profile compliant QIR from a generic unconstrained QIR (if possible).
+
+This document sets out to motivate and demonstrate feasibility of building such a tool.
+
+## Motivation
+
+It is anticipated that most usages of the QIR specificaiton will need to only use a subset of it. These subsets
+may further be subject to constraints such as how one allocate or aquire a qubit handle. We refer subset with
+constraints as a profile. For instance, it is likely that early versions of quantum hardware will have a limited
+set of classical instructions available. With this in mind, the vendor or user of said hardware would define a profile
+that only contains a specified subset. One example of such a profile, is the [base profile](https://github.com/microsoft/qsharp-language/blob/ageller/profile/Specifications/QIR/Base-Profile.md)
+which only allows function calls and branching, but no arithmetic, classical memory or classical registers.
+
+The generation of QIR according to the spec with no constraints would typically be performed by the frontend. A couple
+of examples are Q# or OpenQASM. However, for the generated QIR to be practical it is require to reduce it using a profile
+which is compatible with the target platform:
+
+```text
+┌──────────────────────┐
+│                      │
+│       Frontend       │
+│                      │
+└──────────────────────┘
+           │
+           │
+           ▼
+┌──────────────────────┐
+│                      │
+│         QIR          │
+│                      │
+└──────────────────────┘
+           │
+   QIR Adaptor Tool
+           │
+           ▼
+┌──────────────────────┐
+│                      │
+│     QIR Profile      │
+│                      │
+└──────────────────────┘
+```
+
+We propose that such a reduction could be done using the LLVM passes infrastructure to compose a profile
+which would map the QIR to a subset of the available instructions with any required constraints.
+
+## Feasibility Study
```

Comment:
For a general feasibility study, the two things that would worry me most are 1) that information properly propagates across function boundaries, and 2) how callable values (and captures) can be dealt with. Captured arrays might be interesting. In any case, the work as started makes sense - there are just a bunch of hard cases down the line that need some work to deal with.

Reply:
You will need to expand on this as it is not entirely clear what you mean here with regards to 1. Could you give an example?

For callables, it is unclear whether you mean lambdas or function pointers. Regardless, if the implementation is a first-class LLVM implementation, it should not be difficult to do as you would just inline lambdas the same way you would do with any other function call. The same for captures: These would follow well defined structures that can be parsed. If on the other hand, callables are implemented with opaque objects `Callable`  with custom functions for invoking say `__invoke_callable(%Callable %fnc, %Tuple %args)` it becomes a lot harder since the implementational details depends on the runtime and you cannot provide many guarantees on the correctness.